extensions/markdown-language-features/src/features/smartSelect.ts

```diff
@@ -0,0 +1,153 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Token } from 'markdown-it';
+import * as vscode from 'vscode';
+import { MarkdownEngine } from '../markdownEngine';
+import { TableOfContentsProvider, TocEntry } from '../tableOfContentsProvider';
+
+export default class MarkdownSmartSelect implements vscode.SelectionRangeProvider {
+
+	constructor(
+		private readonly engine: MarkdownEngine
+	) { }
+
+	public provideSelectionRanges(document: vscode.TextDocument, positions: vscode.Position[], _token: vscode.CancellationToken): Promise<vscode.SelectionRange[]> {
+		return Promise.all(positions.map((position) => {
+			return this.provideSelectionRange(document, position, _token);
+		}));
+	}
+
+	private async provideSelectionRange(document: vscode.TextDocument, position: vscode.Position, _token: vscode.CancellationToken): Promise<vscode.SelectionRange> {
+		const headerRange = await this.getHeaderSelectionRange(document, position);
+		const blockRange = await this.getBlockSelectionRange(document, position, headerRange);
+		return blockRange ? blockRange : headerRange ? headerRange : new vscode.SelectionRange(new vscode.Range(new vscode.Position(0, 0), new vscode.Position(0, 0)));
```

Comment:
@mattbierner what else should I return instead of  `: new vscode.SelectionRange(new vscode.Range(new vscode.Position(0, 0), new vscode.Position(0, 0)));`  if both the headerRange and blockRange are undefined?